evals/cybench/dataset.py

```diff
@@ -0,0 +1,150 @@
+from pathlib import Path
+from typing import Callable
+
+import yaml
+
+from inspect_ai.dataset import Dataset, MemoryDataset, Sample
+
+SUPPORTED_FILE_NAMES = ["challenge.yaml", "challenge.yml"]
+
+
+def cybench_dataset(
+    challenges_dir: Path | list[Path] | None,
+    filters: Callable[[Sample], bool] | list[Callable[[Sample], bool]] | None,
+) -> Dataset:
+    # If no path specified, run all the challenges.
+    if not challenges_dir:
+        challenges_dir = Path(__file__).resolve().parent
+    challenges_dir = (
+        challenges_dir if isinstance(challenges_dir, list) else [challenges_dir]
+    )
+    challenges = _get_challenges(challenges_dir)
+    dataset = MemoryDataset(samples=_create_samples(challenges))
+    if filters:
+        dataset = _filter_dataset(dataset, filters)
+    return dataset
+
+
+def _get_challenges(challenges_dirs: list[Path]) -> list[Path]:
+    def get_challenges(challenge_dir: Path) -> list[Path]:
+        if not isinstance(challenge_dir, Path):
+            raise TypeError(
+                "Expected challenge_dir to be a Path, "
+                f"but got {type(challenge_dir).__name__}."
+            )
+
+        challenge_dir = challenge_dir.resolve()
+        if not challenge_dir.exists():
+            raise FileNotFoundError(f"Could not find: {challenge_dir}")
+
+        challenge_dirs = set()
+        for file_name in SUPPORTED_FILE_NAMES:
+            for file in challenge_dir.rglob(file_name):
+                challenge_dirs.add(file.parent)
+        return list(challenge_dirs)
+
+    all_challenge_dirs = set()
+    for challenge_dir in challenges_dirs:
+        all_challenge_dirs.update(get_challenges(challenge_dir))
+
+    return list(all_challenge_dirs)
+
+
+def _create_samples(challenges: list[Path]) -> list[Sample]:
+    def create_sample_per_variant(challenge: Path) -> list[Sample]:
+        challenge_info = _load_challenge_info(challenge)
+        id = challenge_info["name"]
+        target = challenge_info["flag"]
+        challenge_files = _get_files(challenge_info.get("files", {}), challenge)
+        sandbox = _get_sandbox(challenge_info, challenge)
+        tags = challenge_info.get("tags", [])
+        variants = challenge_info["variants"]
+
+        samples = []
+        if isinstance(variants, list):
+            variants = {
+                f"variant_{i}": {"prompt": variant}
+                for i, variant in enumerate(variants)
+            }
+        for name, variant in variants.items():
+            variant_files = _get_files(variant.get("files", {}), challenge)
+            variant_files.update(challenge_files)
+
+            sample = Sample(
+                id=id,
+                input=variant["prompt"],
+                target=target,
+                files=variant_files,
+                metadata={
```

Comment:
Just pulling in the discussion from the PR in the [internal `ctf-evals-template` repo](https://github.com/AI-Safety-Institute/ctf-evals-template/pull/147#discussion_r1768796811) - consider supporting an optional `metadata` field in `challenge.yaml` which is automatically copied into the `Sample` metadata. This way if people want to add arbitrary metadata (e.g. info on the competition or time taken for human to solve) they can just put it under `metadata` rather than needing to also need to update this code?